apic_ml2/neutron/plugins/ml2/drivers/cisco/apic/mechanism_apic.py

```diff
@@ -197,21 +197,61 @@ def __init__(self):
             ofcst.AGENT_TYPE_OPFLEX_OVS)
         ha_ip_db.HAIPOwnerDbMixin.__init__(self)
 
-    def _is_dvs_vif_type(self, context, agent):
-        """Return if this port is a DVS vif
+    def _agent_bind_port(self, context, agent_list):
+        """Attempt port binding per agent.
 
-           We need to bind the port as a DVS VIF type
-           when the port belongs to nova, and when there's
-           an OpFlex agent on that (compute) host that's told
-           us it's supporting a VMware hypervisor.
+           Perform the port binding for a given agent.
+           Returns True if bound successfully.
         """
-        port = context.current
-        if port['device_owner'].startswith('compute:'):
-            hv_type = agent['configurations'].get('hypervisor_type', None)
-            if hv_type and hv_type == acst.HYPERVISOR_VCENTER:
-                return True
+        for agent in agent_list:
+            LOG.debug("Checking agent: %s", agent)
+            if agent['alive']:
+                for segment in context.segments_to_bind:
+                    if self.try_to_bind_segment_for_agent(context, segment,
+                                                          agent):
+                        LOG.debug("Bound using segment: %s", segment)
+                        return True
+            else:
+                LOG.warning(_("Refusing to bind port %(pid)s to dead agent: "
+                              "%(agent)s"),
+                            {'pid': context.current['id'], 'agent': agent})
         return False
 
+    def bind_port(self, context):
+        """Get port binding per host.
+
+           Overriding the superclass implementation
+           in order to support multiple agent types
+           in a single mechanism driver (DVS and OpFlex)
+        """
+
+        LOG.debug("Attempting to bind port %(port)s on "
+                  "network %(network)s",
+                  {'port': context.current['id'],
+                   'network': context.network.current['id']})
+        vnic_type = context.current.get(portbindings.VNIC_TYPE,
+                                        portbindings.VNIC_NORMAL)
+        if vnic_type not in self.supported_vnic_types:
+            LOG.debug("Refusing to bind due to unsupported vnic_type: %s",
+                      vnic_type)
+            return
+
+        # Attempt to bind ports for DVS agents for nova-compute daemons
+        # first. This allows having network agents (dhcp, metadata)
+        # that typically run on a network node using an OpFlex agent to
+        # co-exist with nova-compute daemons for ESX, which host DVS agents.
+        if context.current['device_owner'].startswith('compute:'):
+            agent_list = context.host_agents(acst.AGENT_TYPE_DVS)
+            if self._agent_bind_port(context, agent_list):
```

Comment:
The fact that you are calling _agent_bind_port() separately for each agent type forces the use of if/else statements throughout the following code. I suggest instead implementing separate _bind_dvs_port() and _bind_opflex_port() functions, so these conditionals aren't needed.

Also, the AgentMechanismDriverBase and SimpleAgentMechanismDriverBase classes were structured with specific functions intended to be defined/overridden at each level in the hierarchy. Since you aren't using these, you don't need to use this same structure. I think the code would be easier to read an maintain if the logic from _agent_bind_port(), try_to_bind_segment_for_agent() and _check_segment_for_agent() was all rolled together (without the conditionals) into _bind_dvs_port() and _bind_opflex_port().